include/appbase/application_base.hpp

```diff
@@ -0,0 +1,385 @@
+   }
+   ///@}
+
+   application_base(); ///< protected because application is a singleton that should be accessed via instance()
+
+private:
+   // members are ordered taking into account that the last one is destructed first
+   std::function<void()> sighup_callback;
+   std::function<void()> stop_executor_cb;
+   std::function<void(int, std::function<void()>)> post_cb;
+
+   map<std::type_index, erased_method_ptr> methods;
+   map<std::type_index, erased_channel_ptr> channels;
+
+   std::unique_ptr<class application_impl> my;
+
+   map<string, std::unique_ptr<abstract_plugin>> plugins; ///< all registered plugins
+   vector<abstract_plugin*> initialized_plugins;          ///< stored in the order they were started running
+   vector<abstract_plugin*> running_plugins;              ///< stored in the order they were started running
+
+   inline static std::vector<std::function<void(application_base&)>> plugin_registrations;
+
+   void start_sighup_handler(std::shared_ptr<boost::asio::signal_set> sighup_set);
+   void set_program_options();
+   void write_default_config(const bfs::path& cfg_file);
+   void print_default_config(std::ostream& os);
+
+   void wait_for_signal(std::shared_ptr<boost::asio::signal_set> ss);
+   void setup_signal_handling_on_ios(boost::asio::io_service& ios, bool startup);
+
+   void handle_exception(std::exception_ptr eptr, std::string_view origin);
+};
+
+// ------------------------------------------------------------------------------------------
+template <class executor_t>
+class application_t : public application_base {
+public:
+   static application_t& instance() {
+      if (__builtin_expect(!!app_instance, 1))
+         return *app_instance;
+      app_instance.reset(new application_t);
+      return *app_instance;
+   }
+
+   static void reset_app_singleton() {
+      app_instance.reset();
+   }
+
+   static bool null_app_singleton() {
+      return !app_instance;
+   }
+
+   /**
+    * Post func to run on io_service with given priority.
+    *
+    * -- deprecated: use app().executor().post()
+    *
+    * @param priority can be appbase::priority::* constants or any int, larger ints run first
+    * @param func function to run on io_service
+    * @return result of boost::asio::post
+    */
+   template <typename Func>
+   auto post(int priority, Func&& func) {
+      return executor_.post(priority, std::forward<Func>(func));
+   }
+
+   /**
+    *  Wait until quit(), SIGINT or SIGTERM and then shutdown.
+    *  Should only be executed from one thread.
+    */
+   void exec() {
+      application_base::exec(executor_);
+   }
+
+   boost::asio::io_service& get_io_service() {
+      return executor_.get_io_service();
+   }
+
+   void startup() {
+      application_base::startup(get_io_service());
+   }
+
+   application_t() {
+      set_stop_executor_cb([&]() { get_io_service().stop(); });
+      set_post_cb([&](int prio, std::function<void()> cb) { executor_.post(prio, std::move(cb)); });
```

Comment:
I don't see what this buys you over just using `executor_.post()` since you are hard-coding it here. It seems the only way this could be made to work generically is if you registered this callbacks before construction.

Reply:
That's because `application_base` uses `post()` and the executor is not defined yet, so we can't use `executor_` from application_base.

Reply:
> Ah, yes, you changed where executor_ is stored.

Not since the PR was made, but yes before there was one class for `application`, now there is a base which has its regular members implemented in a .cpp file, and the `application` class itself which is a template class in a header.